src/pubsub/blabber.rs

```diff
@@ -0,0 +1,172 @@
+    let validator = MessageValidator::new();
+
+    let connector = Connector::new(&config)?;
+
+    // this unwrap will succeed because we wouldn't be creating these tasks if
+    // there wasn't a client config.
+    let pubsub_config = config.pubsub().unwrap();
+
+    let mut stream = None;
+    let mut read_buffer = Buffer::new(pubsub_config.read_buffer_size());
+
+    while RUNNING.load(Ordering::Relaxed) {
+        if stream.is_none() {
+            CONNECT.increment();
+            PUBSUB_SUBSCRIBE.increment();
+
+            stream = match timeout(
+                pubsub_config.connect_timeout(),
+                connector.connect(&endpoint),
+            )
+            .await
+            {
+                Ok(Ok(s)) => {
+                    CONNECT_OK.increment();
+                    CONNECT_CURR.increment();
+                    PUBSUB_SUBSCRIBER_CURR.add(1);
+                    PUBSUB_SUBSCRIBE_OK.increment();
+
+                    Some(s)
+                }
+                Ok(Err(_)) => {
+                    CONNECT_EX.increment();
+                    PUBSUB_SUBSCRIBE_EX.increment();
+
+                    sleep(Duration::from_millis(100)).await;
+                    continue;
+                }
+                Err(_) => {
+                    CONNECT_TIMEOUT.increment();
+                    PUBSUB_SUBSCRIBE_EX.increment();
+
+                    sleep(Duration::from_millis(100)).await;
+                    continue;
+                }
+            }
+        }
+
+        let mut s = stream.take().unwrap();
+
+        // read until response or timeout
+        loop {
+            match s.read(read_buffer.borrow_mut()).await {
+                Ok(n) => {
+                    unsafe {
+                        read_buffer.advance_mut(n);
+                    }
+                    {
+                        loop {
+                            let consumed = {
+                                let rbuf: &[u8] = read_buffer.borrow();
+
+                                if rbuf.len() >= HEADER_LEN as usize {
+                                    let len =
+                                        u32::from_be_bytes(rbuf[0..4].try_into().unwrap()) as usize;
+
+                                    // check if we have only a partial message
+                                    if rbuf.len() < len {
+                                        break;
+                                    }
+
+                                    let mut mbuf = rbuf[8..len].to_owned();
+
+                                    let _ = validator.validate(&mut mbuf);
+
+                                    // return the number of bytes consumed
+                                    len
+                                } else {
+                                    break;
+                                }
+                            };
+
+                            read_buffer.advance(consumed);
```

Comment:
Should we have a `PUBSUB_RECEIVE.increment()` and a `PUBSUB_RECEIVE_OK.increment()` here?

Reply:
This happens up one level in the MessageValidator.

Reply:
Ack, thanks.